### PR TITLE
Add an MLEM solver

### PR DESCRIPTION
**Pull request by lesaintjerome**
## Description

Standard MLEM solver.

## TODO

- [x] Implement new feature.
- [x] Add or update unittests.
- [x] Add or update docstrings.

### CODE REVIEW

File: corrct/_projector_backends.py [hunk 1]
```diff
@@ -487,6 +487,9 @@ def initialize_geometry(
 
             if prj_geom.det_shape_vu is None:
                 prj_geom.det_shape_vu = np.array(self.vol_geom.shape_xyz[list([0])], dtype=int)
+            else:
+                self.prj_shape_vwu = np.array([num_angles, prj_geom.det_shape_vu[0]])
```
Comment (Obi-Wan) [inline corrct/_projector_backends.py]: This chunk looks unrelated to the subject of the commit. Moreover, I'm not sure why it is needed

File: examples/example_10_synthetic_phantom_dead-zone_MLEM_vs_SIRT.py [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+ph_or = ph_or[:, :, 1]
+
+(ph, vol_att_in, vol_att_out) = cct.testing.phantom_assign_concentration(ph_or)
+(sino, angles, expected_ph, background_avg) = cct.testing.create_sino(
+    ph, 30, psf=None, add_poisson=True, dwell_time_s=1e-2, background_avg=1e-2, background_std=1e-4
+)
+
+bckgnd_weight = np.sqrt(background_avg / (vol_shape[0] * np.sqrt(2)))
+
+num_iterations_sirt = 10000
```
Comment (Obi-Wan) [inline examples/example_10_synthetic_phantom_dead-zone_MLEM_vs_SIRT.py]: This is a rather odd number for SIRT iterations. Maybe rather 100?

**Comment by lesaintjerome** [comment]
CI failed due to multiplication by `x_mask` without checking presence of it. Corrected in [b8dd556](https://github.com/cicwi/PyCorrectedEmissionCT/pull/20/commits/b8dd55652d8efdb348887100fc32bd791195585b)

File: corrct/solvers.py [hunk 3]
```diff
@@ -551,6 +551,164 @@ def __call__(  # noqa: C901
...
+    data_term_test : Optional[DataFidelityBase], optional
+        The data fidelity to be used for the test set.
+        If None, it will use the same as for the rest of the data.
+        The default is None.
+    """
+
+    def __init__(
+        self,
+        verbose: bool = False,
+        relaxation: float = 1.95,
```
Comment (Obi-Wan) [inline corrct/solvers.py]: Is this value correct or even used?

File: corrct/solvers.py [hunk 4]
```diff
@@ -551,6 +551,164 @@ def __call__(  # noqa: C901
...
+        tolerance: Optional[float] = None,
+        regularizer: Union[Sequence[BaseRegularizer], BaseRegularizer, None] = None,
+        data_term: Union[str, DataFidelityBase] = "l2",
```
Comment (Obi-Wan) [inline corrct/solvers.py]: This might also not be needed, as the noise model of MLEM is fixed as Poisson -> data term = KL. Should it be ignored or maybe made optional?

File: corrct/solvers.py [hunk 5]
```diff
@@ -551,6 +551,164 @@ def __call__(  # noqa: C901
...
+        """
+        Return the MLEM info.
+
+        Returns
+        -------
+        str
+                 info string.
+        """
+        reg_info = "".join(["-" + r.info().upper() for r in self.regularizer])
+        return Solver.info(self) + "-" + self.data_term.info() + reg_info
```
Comment (Obi-Wan) [inline corrct/solvers.py]: This might lead to incorrect info, as the noise model is fixed and regularization should not be supported (at least for this first implementation iteration)

File: corrct/solvers.py [hunk 6]
```diff
@@ -551,6 +551,164 @@ def __call__(  # noqa: C901
...
+        A: operators.BaseTransform,
+        b: NDArrayFloat,
+        iterations: int,
+        x0: Optional[NDArrayFloat] = None,
+        lower_limit: Union[float, NDArrayFloat, None] = None,
+        upper_limit: Union[float, NDArrayFloat, None] = None,
+        x_mask: Optional[NDArrayFloat] = None,
+        b_mask: Optional[NDArrayFloat] = None,
+        b_test_mask: Optional[NDArrayFloat] = None,
+        mu: float = 1e-5,
```
Comment (Obi-Wan) [inline corrct/solvers.py]: Where is this parameter `mu` used? Could it be a leftover from a previous implementation?

File: examples/example_10_synthetic_phantom_dead-zone_MLEM_vs_SIRT.py [hunk 7]
```diff
@@ -0,0 +1,147 @@
+# -*- coding: utf-8 -*-
+"""
+This example reproduces the l2 dead-zone reconstruction of the phantom used in:
+
+- N. Viganò and V. A. Solé, “Physically corrected forward operators for
+induced emission tomography: a simulation study,” Meas. Sci. Technol., no.
+Advanced X-Ray Tomography, pp. 1–26, Nov. 2017.
+
+@author: Nicola VIGANÒ, Computational Imaging group, CWI, The Netherlands,
```
Comment (Obi-Wan) [inline examples/example_10_synthetic_phantom_dead-zone_MLEM_vs_SIRT.py]: Author should be adjusted, as I'm not the author of this example :)

File: examples/example_10_synthetic_phantom_dead-zone_MLEM_vs_SIRT.py [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+
+lowlim_l2 = cct.solvers.Constraint_LowerLimit(0, norm=cct.solvers.DataFidelity_l2())
+lowlim_l2w = cct.solvers.Constraint_LowerLimit(0, norm=cct.solvers.DataFidelity_wl2(1 / bckgnd_weight))
+
+data_term_ls = cct.solvers.DataFidelity_l2()
+data_term_lsw = cct.solvers.DataFidelity_wl2(sino_weights)
+data_term_lsb = cct.solvers.DataFidelity_l2b(sino_variance)
+
+regul_weight = 1
+regul = cct.regularizers.Regularizer_TV2D(regul_weight)
```
Comment (Obi-Wan) [inline examples/example_10_synthetic_phantom_dead-zone_MLEM_vs_SIRT.py]: Is this a left-over?

File: examples/example_10_synthetic_phantom_dead-zone_MLEM_vs_SIRT.py [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+with cct.projectors.ProjectorUncorrected(ph.shape, angles) as A:
+
+    solver_wls_l = cct.solvers.PDHG(verbose=True, data_term=data_term_lsw, regularizer=[lowlim_l2w])
+    rec_wls_l, _ = solver_wls_l(A, sino_substract, num_iterations_sirt, x_mask=vol_mask)
+    #
+    solver_sirt = cct.solvers.SIRT(verbose=True, data_term=data_term_lsw)
+    rec_sirt, _ = solver_sirt(A, sino_substract, num_iterations_sirt, x_mask=vol_mask)
+
+    solver_mlem = cct.solvers.MLEM(verbose=True)
+    rec_mlem, _ = solver_mlem(A, sino_substract, num_iterations_mlem, x_mask=vol_mask, lower_limit=1e-5, upper_limit=10)
```
Comment (Obi-Wan) [inline examples/example_10_synthetic_phantom_dead-zone_MLEM_vs_SIRT.py]: why is there an upper limit? Do we have some slight instabilities?
If yes, I suggest to have a look at some other implementation (e.g. the one from @astra-toolbox, to see how they handle it)

**Review by Obi-Wan — COMMENTED** [review]
Overall very good addition. There might be some minor points to fix. Thanks!

File: corrct/_projector_backends.py (hunk 1 above)
Comment (lesaintjerome) [reply corrct/_projector_backends.py]: OK removed. Did not see that it was set in the mother class.

File: corrct/solvers.py (hunk 3 above)
Comment (lesaintjerome) [reply corrct/solvers.py]: OK. Relaxation unused and removed from the class.

File: corrct/solvers.py (hunk 5 above)
Comment (lesaintjerome) [reply corrct/solvers.py]: Right. Modified to return only `MLEM`.

File: corrct/solvers.py (hunk 6 above)
Comment (lesaintjerome) [reply corrct/solvers.py]: Nowhere (only in the regularized version). But useless here.

File: examples/example_10_synthetic_phantom_dead-zone_MLEM_vs_SIRT.py (hunk 7 above)
Comment (lesaintjerome) [reply examples/example_10_synthetic_phantom_dead-zone_MLEM_vs_SIRT.py]: Modified.

File: examples/example_10_synthetic_phantom_dead-zone_MLEM_vs_SIRT.py (hunk 2 above)
Comment (lesaintjerome) [reply examples/example_10_synthetic_phantom_dead-zone_MLEM_vs_SIRT.py]: That was to check convergence. Reset to 100.

File: examples/example_10_synthetic_phantom_dead-zone_MLEM_vs_SIRT.py (hunk 8 above)
Comment (lesaintjerome) [reply examples/example_10_synthetic_phantom_dead-zone_MLEM_vs_SIRT.py]: Indeed. Removed.

File: examples/example_10_synthetic_phantom_dead-zone_MLEM_vs_SIRT.py (hunk 9 above)
Comment (lesaintjerome) [reply examples/example_10_synthetic_phantom_dead-zone_MLEM_vs_SIRT.py]: From the simple tests I ran, if I remove the upper limit, some voxels (mainly far from the center but inside the fov) explode. Inducing severe streaks artefacts in the reconstruction.
image
Working on it and having a look at what is implemented in ASTRA. Not sure any constraint is implemented. TBC...

File: corrct/solvers.py (hunk 4 above)
Comment (lesaintjerome) [reply corrct/solvers.py]: It is not used. `None` could be cleaner. Being unfamiliar with heritance, a bit unsure of what would be the side effects of removing this parameter from the API...

File: examples/example_10_synthetic_phantom_dead-zone_MLEM_vs_SIRT.py (hunk 9 above)
Comment (Obi-Wan) [reply examples/example_10_synthetic_phantom_dead-zone_MLEM_vs_SIRT.py]: I might have a hint for this, or at least something to make sure doesn't interfere with the reconstruction: The signal is background subtracted, but the information about the background is not included in the MLEM reconstruction.

I mean that the signal `yb` that is passed to the reconstruction, is actually `yb = P(y + b) - b`, where `b` is the background and `P(.)` is the Poisson statistical distribution function. MLEM might be expecting `b = 0`, and thus to have `P(y)` as input data. To be verified!

Comment (lesaintjerome) [reply examples/example_10_synthetic_phantom_dead-zone_MLEM_vs_SIRT.py]: indeed, running the simulation without background noise (`background_avg=0.` and `background_std=0.`) produces "normal" results. Masking or not does not impact the reconstruction in the particular test.
I remove the noise from the test.

File: corrct/solvers.py (hunk 4 above)
Comment (Obi-Wan) [reply corrct/solvers.py]: This has been fixed in the latest commits. KL is now passed by default. Moreover, the background is now honored.

File: examples/example_10_synthetic_phantom_dead-zone_MLEM_vs_SIRT.py (hunk 9 above)
Comment (Obi-Wan) [reply examples/example_10_synthetic_phantom_dead-zone_MLEM_vs_SIRT.py]: The background is now handled correctly
